File: library/src/main/java/org/xmtp/android/library/codecs/ContentCodec.kt
package org.xmtp.android.library.codecs

import com.google.protobuf.kotlin.toByteString
import org.bouncycastle.asn1.cms.CMSAttributes.contentType
import org.xmtp.android.library.Client
import org.xmtp.android.library.EncodedContentCompression
import org.xmtp.proto.message.contents.Content
import uniffi.xmtpv3.FfiEncodedContent

typealias EncodedContent = Content.EncodedContent

fun <T> EncodedContent.decoded(): T? {
    val codec = Client.codecRegistry.find(type)
    var encodedContent = this
    if (hasCompression()) {
        encodedContent = decompressContent()
    }
    return codec.decode(content = encodedContent) as? T
}

fun EncodedContent.compress(compression: EncodedContentCompression): EncodedContent {
    val copy = this.toBuilder()
    when (compression) {
        EncodedContentCompression.DEFLATE -> {
            copy.also {
                it.compression = Content.Compression.COMPRESSION_DEFLATE
            }
        }

        EncodedContentCompression.GZIP -> {
            copy.also {
                it.compression = Content.Compression.COMPRESSION_GZIP
            }
        }
    }
    copy.also {
        it.content = compression.compress(content.toByteArray())?.toByteString()
    }
    return copy.build()
}

fun EncodedContent.decompressContent(): EncodedContent {
    if (!hasCompression()) {
        return this
    }
    var copy = this
    when (compression) {
        Content.Compression.COMPRESSION_DEFLATE -> {
            copy =
                copy
                    .toBuilder()
                    .also {
                        it.content =
                            EncodedContentCompression.DEFLATE
                                .decompress(content = content.toByteArray())
                                ?.toByteString()
                    }.build()
        }

        Content.Compression.COMPRESSION_GZIP -> {
            copy =
                copy
                    .toBuilder()
                    .also {
                        it.content =
                            EncodedContentCompression.GZIP
                                .decompress(content = content.toByteArray())
                                ?.toByteString()
                    }.build()
        }

        else -> return copy
    }
    return copy
}

fun encodedContentFromFfi(ffi: FfiEncodedContent): EncodedContent =
    EncodedContent
        .newBuilder()
        .also { builder ->
            ffi.typeId?.let {
                builder.type = ContentTypeIdBuilder.fromFfi(it)
            }
            builder.putAllParameters(ffi.parameters)
            ffi.fallback?.let {
                builder.fallback = it
            }
            ffi.compression?.let {
                builder.compressionValue = it
            }
            builder.content = ffi.content.toByteString()
        }.build()

interface ContentCodec<T> {
    val contentType: ContentTypeId

    fun encode(content: T): EncodedContent

    fun decode(content: EncodedContent): T

    fun fallback(content: T): String?

    fun shouldPush(content: T): Boolean
}

val id: String
    get() = contentType.id

File: Sources/XMTPiOS/Codecs/ContentCodec.swift
//
//  ContentCodec.swift
//
//
//  Created by Pat Nakajima on 11/28/22.
//

import Foundation

enum CodecError: String, Error {
	case invalidContent, codecNotFound
}

public typealias EncodedContent = Xmtp_MessageContents_EncodedContent

extension EncodedContent {
	public func decoded<T>() throws -> T {
		let codec = Client.codecRegistry.find(for: type)

		var encodedContent = self

		if hasCompression {
			encodedContent = try decompressContent()
		}

		if let content = try codec.decode(content: encodedContent) as? T {
			return content
		}

		throw CodecError.invalidContent
	}

	func compress(_ compression: EncodedContentCompression) throws -> EncodedContent {
		var copy = self

		switch compression {
		case .deflate:
			copy.compression = .deflate
		case .gzip:
			copy.compression = .gzip
		}

		if let compressedContent = compression.compress(content: content) {
			copy.content = compressedContent
		} else {
			throw CodecError.invalidContent
		}

		return copy
	}

	func decompressContent() throws -> EncodedContent {
		if !hasCompression {
			return self
		}

		var copy = self

		switch compression {
		case .gzip:
			if let decompressedContent = EncodedContentCompression.gzip.decompress(content: content) {
				copy.content = decompressedContent
			} else {
				throw CodecError.invalidContent
			}
		case .deflate:
			if let decompressedContent = EncodedContentCompression.deflate.decompress(content: content) {
				copy.content = decompressedContent
			} else {
				throw CodecError.invalidContent
			}
		default:
			return copy
		}

		return copy
	}

}

public protocol ContentCodec: Hashable, Equatable {
	associatedtype T

	var contentType: ContentTypeID { get }
	func encode(content: T) throws -> EncodedContent
	func decode(content: EncodedContent) throws -> T
	func fallback(content: T) throws -> String?
	func shouldPush(content: T) throws -> Bool
}

public extension ContentCodec {
	static func == (lhs: Self, rhs: Self) -> Bool {
		return lhs.contentType.authorityID == rhs.contentType.authorityID && lhs.contentType.typeID == rhs.contentType.typeID
	}

	var id: String {
		contentType.id
	}

	func hash(into hasher: inout Hasher) {
		hasher.combine(id)
	}

	var description: String {
		contentType.description
	}
}
